### PR TITLE
Add Documentation on Reference Interpreter

### LINKED ISSUE #963: Add interpreter for BatchNormInferenceOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1121: Add interpreter for BatchNormGradOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1122: Add interpreter for BatchNormTrainingOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1123: Add interpreter for CholeskyOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1138: Add interpreter for TriangularSolveOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1225: Enable interpretation of dot
Lower DotOp to DotGeneralOp, make sure that the interpreting all the operations in the result is implemented.

### LINKED ISSUE #1226: Enable interpretation of einsum
Lower EinsumOp to other ops, make sure that the interpreting all the operations in the result is implemented.

### LINKED ISSUE #1227: Enable interpretation of torch_index_select
Lower TorchIndexSelectOp to GatherOp, make sure that the interpreting all the operations in the result is implemented.

### LINKED ISSUE #1799: Enable interpretation of broadcast

### LINKED ISSUE #1800: Enable interpretation of create_token

### LINKED ISSUE #1801: Enable interpretation of cross-replica-sum

### LINKED ISSUE #1802: Enable interpretation of unary_einsum

### PR DESCRIPTION
**Pull request by ghpvnist**
This PR adds documentation to the StableHLO interpreter's current status and provides a way to convert StableHLO ops without a direct implementation to ones that do via running a set of expander/transform passes within the [openxla/xla](https://github.com/openxla/xla) repo.

In summary, this PR closes most of the remaining specced ops without an interpreter with a few exceptions noted in the doc: `FftOp`, `RngOp`, `RngBitGeneratorOp`, `UniformDequantizeOp`, and `UniformQuantizeOp`. It also documents interpretation of ops categorized as "Not In HLO" which currently does not have a formal spec in our StableHLO [spec.md](https://github.com/openxla/stablehlo/blob/main/docs/spec.md).

This PR resolves most of the items in #1571 except ones mentioned above and those marked N/A in the comment (rationale for these are provided in the doc).

Miscellaneous:
closes #1121 batch_norm_grad
closes #963 batch_norm_inference
closes #1122 batch_norm_training
closes #1123 cholesky
closes #1138 triangular_solve

Not In HLO:
closes #1799 broadcast
closes #1800 create_token
closes #1801 cross-replica-sum
closes #1225 dot
closes #1226 einsum
closes #1227 torch_index_select
closes #1802 unary_einsum

### CODE REVIEW

File: docs/interpreter_status.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+The main goal of the StableHLO interpreter is to provide a reference
+implementation to the semantics of StableHLO opset according to its
+specification. The secondary goal is for the implementation to closely follow
+the spec to provide additional clarity to the semantics of even the most
+involved operations like `Convolution`, `Gather`/`Scatter`, and `DotGeneral`.
+
+At the moment, OpenXLA supports the interpretation of 91 out of 96 specced
+StableHLO ops, with the remaining 5 ops (`FftOp`, `RngOp`, `RngBitGeneratorOp`,
+`UniformDequantizeOp`, `UniformQuantizeOp`) being a work in progress (see
```
Comment (GleasonK) [inline docs/interpreter_status.md:11]: is "work in progress" accurate?

Perhaps "initial investigation completed and documented in <link tickets>, these final enhancements will be implemented on an as-needed community basis"

File: docs/interpreter_status.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
+# StableHLO Interpreter
+
+The main goal of the StableHLO interpreter is to provide a reference
+implementation to the semantics of StableHLO opset according to its
+specification. The secondary goal is for the implementation to closely follow
+the spec to provide additional clarity to the semantics of even the most
```
Comment (GleasonK) [inline docs/interpreter_status.md:6]: "The secondary goal is for the implementation to closely follow
the spec, favoring readability over performance, to provide additional..."

I was unclear on initial read what this meant, but I think you meant that the impls are intended to be readable?

File: docs/interpreter_status.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+`UniformDequantizeOp`, `UniformQuantizeOp`) being a work in progress (see
+[status.md](https://github.com/openxla/stablehlo/blob/main/docs/status.md) for
+a complete list of ops and its latest status).
+
+## Scope
+
+We categorized the StableHLO opset into 11 categories consisting of 118 ops in
+total (see [Appendix](#appendix)). According to our
+[roadmap](https://github.com/openxla/stablehlo/blob/main/docs/roadmap.md), one
+goal of StableHLO v1.0 is to implement a reference interpreter for all specced
```
Comment (GleasonK) [inline docs/interpreter_status.md:20]: The actual wording is "We are planning to complete all or almost all work in this workstream in StableHLO v1.0."

Perhaps use the phrasing "all or most all specced ops" here to match this

File: docs/interpreter_status.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+<!-- markdownlint-enable line-length -->
+
+### Quantization
+
+There are two specced ops in this category that the interpreter does not support
+at the moment. These ops are part of an ongoing Quantization work. Supporting
+interpretation of these two ops is a WIP:
+
+* `UniformDequantizeOp`
+* `UniformQuantizeOp`
```
Comment (GleasonK) [inline docs/interpreter_status.md:106]: And uniform quant type. Do we have a ticket for this work?

I'd make this section simply say:
"Quantization spec'ing is still WIP, and interpreter support for quantized ops and types is tracked by ticket #????"

That will keep it from going out of date.

File: docs/interpreter_status.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+## Build and Run the Reference Interpreter
```
Comment (GleasonK) [inline docs/interpreter_status.md:108]: ## Usage Instructions

### Building the Reference Interpreter

...

### The Interpreter Dialect

... Brief description and link to TD file

### The Check Dialect

... Brief description and link to TD file

### Writing test programs

...

File: docs/interpreter_status.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+<!-- markdownlint-disable line-length -->
+We use LLVM's [lit](https://llvm.org/docs/CommandGuide/lit.html) tool to run,
+and compare against expected value to test the interpreter (see [stablehlo/tests](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests)
+for sample tests).
+<!-- markdownlint-enable line-length -->
+
+A custom `Check` dialect is used to compare interpreter runtime values with
+expected values. The `Check` ops are embedded into the StableHLO programs which
+is run via callbacks through `stablehlo.custom_call`.
```
Comment (GleasonK) [inline docs/interpreter_status.md:129]: Lets leave off custom_call, this is a custom solution for when tests need to be serialized which is out of scope of this doc.

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: docs/interpreter_status.md (hunk 3 above)
Comment (ghpvnist) [reply docs/interpreter_status.md]: Good point, instead of reinventing the wording, I directly copied the wording for consistency.

File: docs/interpreter_status.md (hunk 5 above)
Comment (ghpvnist) [reply docs/interpreter_status.md]: This is a good idea, I've updated the doc, PTAL.

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: docs/interpreter_status.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+v1.0. Of the 96 ops that have a spec currently, we can interpret 91 ops through
+OpenXLA (see [Special Cases](#special-cases) for the remaining 5).
+
+## Specification
+
+The main requirement for the interpreter is to have 1:1 correspondence with the
+spec. The spec allows standardization of the interpreter across similar ops that
+lead to modular, high quality implementation of the interpreter.
+
+## Verification
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply docs/interpreter_status.md]: Sure, they were there just to explain the interpreter workstream but may not be relevant to the users of the interpreter.

File: docs/interpreter_status.md [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+op. Supporting interpretation of these two ops is a WIP.
+
+<!-- markdownlint-disable line-length -->
+The tool to convert remaining ops in this category to StableHLO ops that the
+interpreter supports resides in [hlo_expand_main.cc](https://github.com/openxla/xla/blob/main/xla/tools/hlo_expand_main.cc).
+<!-- markdownlint-enable line-length -->
+
+### Not in HLO
+
+Apart from the specced ops, this category consists of 10 unspecced ops (as they
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply docs/interpreter_status.md]: Updated to link to the table below for better access.
